dbms/src/Storages/S3/FileCache.h

```diff
@@ -228,7 +230,19 @@ class FileCache
                                                                              : nullptr;
     }
 
-    static void shutdown() { global_file_cache_instance = nullptr; }
+    // `bgDownload` will add task which holds weak_ptr of `global_file_cache_instance` to S3FileCachePool.
+    // When the task is scheduled in another thread, it will try to lock the weak_ptr to get the shared_ptr.
+    // If the global_file_cache_instance is destructed before the task is scheduled, the weak_ptr will be expired.
+    // If there are still tasks running, the `global_file_cache_instance` should wait until all tasks are finished.
+    static void shutdown()
+    {
+        global_file_cache_initialized.store(false, std::memory_order_release);
+        while (global_file_cache_instance.use_count() > 1)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
```

Comment:
What about add a `S3FileCachePool::shutdown()` here instead if making `global_file_cache_instance` a shared_ptr? That shutdown should wait until all the job in `S3FileCachePool` finished.
https://github.com/pingcap/tiflash/blob/3cda2f61923f47c89db026bd118eded90e76c625/dbms/src/Common/UniThreadPool.cpp#L236-L260

Reply:
fixed